### ISSUE TITLE
Wrong warning on shadowed declaration 

### ISSUE THREAD

**Issue by 4000D** [ethereum/solidity#4558]
Solidity prints the declaration shadowed warning on a contract like 
```solidity
contract A {
  function foo() external { }
  function bar() external {
    uint foo = 10; // warning that internal variable `foo` shadows the function `foo`.
  }
}
```

But actually, because function `foo` is external, it is not internally accessible in `bar` function unless `this.foo()` or `a.foo()` is used, so I think function `foo` is not shadowed by internal variable `foo`.

**Comment by axic** [comment]
Fair point, we may have to discuss this in #3729, but in general I would agree it shouldn't give a false warning.

**Comment by chriseth** [comment]
The call to `resolveName(...)` in `DeclarationRegistrationHelper::registerDeclaration(` - line 484 also searches for invisible declarations - and the function `foo` is invisible. I'm not 100% sure why we did this, but I think it is better to warn about such cases than not warn about other cases.

**Comment by axic** [comment]
Current behaviour - it still exists as a problem:
```
4558.sol:4:5: Warning: This declaration shadows an existing declaration.
    uint foo = 10; // warning that internal variable `foo` shadows the function `foo`.
    ^------^
4558.sol:2:3: The shadowed declaration is here:
  function foo() external { }
  ^-------------------------^
4558.sol:4:5: Warning: Unused local variable.
    uint foo = 10; // warning that internal variable `foo` shadows the function `foo`.
    ^------^
4558.sol:3:3: Warning: Function state mutability can be restricted to pure
  function bar() external {
  ^ (Relevant source part starts here and spans across multiple lines).
```

**Comment by Marenz** [comment]
Is this ticket blocked by https://github.com/ethereum/solidity/pull/3729 or is there a concrete next step that can be taken to close this?

**Comment by chriseth** [comment]
Not sure - what do you think, @axic?

**Comment by axic** [comment]
You are right, I think it would simplify the situation.  It is a low-priority issue so I would suggest to perhaps postpone it until that is solved.

**Comment by chriseth** [comment]
Actually it might be a good idea to keep the warning (and just change the text) - it is still easy to confuse the variable and the function even if it would not be visible.

**Comment by chriseth** [comment]
Related to https://github.com/ethereum/solidity/issues/4869

**Comment by axic** [comment]
On the meeting there were no strong opinions whether to keep this warning or not, but we were leaning towards keeping it because missing the fact it is `external` and in fact not actual shadowing is easy, thus the warning is helpful to avoid mistakes.

There was agreement though to improving the message would be nice.
